src/NHibernate/Linq/ReWriters/AddJoinsReWriter.cs

```diff
@@ -59,6 +62,29 @@ public override void VisitNhHavingClause(NhHavingClause havingClause, QueryModel
 			_whereJoinDetector.Transform(havingClause);
 		}
 
+		public override void VisitJoinClause(JoinClause joinClause, QueryModel queryModel, int index)
+		{
+			VisitJoinClause(joinClause, queryModel, joinClause);
+		}
+
+		private void VisitJoinClause(JoinClause joinClause, QueryModel queryModel, IBodyClause bodyClause)
+		{
+			joinClause.InnerSequence = _whereJoinDetector.Transform(joinClause.InnerSequence);
+
+			// When associations are located in the outer key (e.g. from a in A join b in B b on a.C.D.Id equals b.Id),
+			// we have to insert the association join before the current join in order to produce a valid query.
```

Comment:
Does this mean that such queries supported in LINQ but in HQL similar queries will fail:
`from A a inner join B b with a.C.D.Id = b.Id`
`from A a inner join B b with b.C.D.Id = a.Id`

Reply:
> from A a inner join B b with a.C.D.Id = b.Id

This case is supported by HQL and for each association an inner join is added where in this PR a left outer join is added instead. After a second thought about what type of join would need to be used, I removed the logic for the outer key and leave it to HQL for adding inner joins. I had the following query in mind: 
```c#
from A a inner join B b with a.C.D.NullableProp = b.NullableProp
```
Linq to Objects would throw a NRE when `a.C` or `a.C.D` is null, but what should we do when generating the query? There are two options:
- Use inner joins for associations as HQL does today
- Use left joins for associations and use the logic that was done in #1996 in order to simulate what Linq to Objects would do for: `a?.C?.D?.NullableProp == b?.NullableProp`

My conclusion was that the second option is not practical in real world scenarios as it would join rows with `null` values.

> from A a inner join B b with b.C.D.Id = a.Id

This case is not supported by HQL.

Reply:
> would join rows with null values.

But NULL values won't be joined anyway in DB unless you explicitly add `is null` check. null = null evaluates to false. So yes - left join seems unnecessary.

Reply:
> But NULL values won't be joined anyway in DB unless you explicitly add is null check

By using the second option, the logic from #1996 would add the null checks

Reply:
The last commit switches left to inner join also for `from A a inner join B b with b.C.D.Id = a.Id` case.

Reply:
> `from A a inner join B b with b.C.D.Id = a.Id`
> This case is not supported by HQL

Ideally this case should be fixed in HQL but I guess it's out of scope for this PR.

Regarding switch to inner join for associations. What about left join?
`from A a left join B b with b.C.D.Id = a.Id`

That seems to be the case when left joins need to be added for associations. Again this PR is not about left joins but could you please clarify how it will work when #2328 is merged?

Reply:
> could you please clarify how it will work when #2328 is merged?

The [last commit](https://github.com/nhibernate/nhibernate-core/pull/2328/commits/777701eaa268c173f2d4b730766baa37022cf0a7) in #2328 addresses the issue that the inner join is added only when the "parent join" is an inner join. By defualt `_whereJoinDetector` add joins as a left outer join, so we have to use `joinClause.MakeInner()` when we want an inner join. In #2328 `NhOuterJoinClause` class was added in order to distinguish whether is an outer join or not.

Reply:
Got it. Thanks